swarm/src/handler.rs

```diff
@@ -158,6 +158,17 @@ pub trait ConnectionHandler: Send + 'static {
         >,
     >;
 
+    /// Gracefully close the [`ConnectionHandler`].
+    ///
+    /// The contract for this function is equivalent to a [`Stream`](futures::Stream).
+    /// It will be polled to completion _in parallel_ with closing the underlying physical connection.
+    /// Performing IO on the connection will likely fail.
```

Comment:
Why not first poll `ConnectionHandler::poll_close` to completion and then close the underlying physical connection?

All components in a `Swarm` including the various `ConnectionHandler`s are expected to be cooperative, e.g. eventually return `Poll::Ready(None)`. Why rush it, thus taking away the opportunity of `ConnectionHandler` to e.g. correctly close its channels?

This would also be in line with `Swarm::close_connection` being _graceful_.

https://github.com/libp2p/rust-libp2p/blob/1e3ffeb7990e23660e30dd95fbe48c1a7138f102/swarm/src/lib.rs#L637-L646

Reply:
Yeah that is a good idea! :)